plugins/org.yakindu.sct.model.stext/src/org/yakindu/sct/model/stext/scoping/ImportUriProvider.java

```diff
@@ -0,0 +1,114 @@
+import org.yakindu.sct.model.stext.extensions.STextExtensions;
+import org.yakindu.sct.model.stext.scoping.IPackageImport2URIMapper.PackageImport;
+import org.yakindu.sct.model.stext.stext.ImportScope;
+import org.yakindu.sct.model.stext.stext.StatechartSpecification;
+import org.yakindu.sct.model.stext.stext.StextPackage;
+
+import com.google.inject.Inject;
+
+/**
+ * @author Thomas Kutz - Extracted from {@link STextGlobalScopeProvider}
+ */
+public class ImportUriProvider {
+
+	@Inject
+	private IPackageImport2URIMapper mapper;
+	@Inject
+	private STextExtensions utils;
+	
+	public LinkedHashSet<URI> get(Resource resource) {
+		final LinkedHashSet<URI> uniqueImportURIs = new LinkedHashSet<>(5);
+		IAcceptor<String> collector = new URICollector(resource.getResourceSet(), uniqueImportURIs);
+		
+		collectFromImportScopes(resource, collector);
+		collectImplicitImports(resource, collector);
+		
+		removeInvalidUris(resource, uniqueImportURIs);
+		return uniqueImportURIs;
+	}
+
+	protected void removeInvalidUris(Resource resource, final LinkedHashSet<URI> uniqueImportURIs) {
+		Iterator<URI> uriIter = uniqueImportURIs.iterator();
+		while (uriIter.hasNext()) {
+			if (!isValidUri(resource, uriIter.next().trimQuery()))
+				uriIter.remove();
+		}
+	}
+
+	protected void collectFromImportScopes(Resource resource, IAcceptor<String> collector) {
+		Collection<ImportScope> importScopes = getImportScopes(resource);
+		for (ImportScope object : importScopes) {
+			EList<String> imports = object.getImports();
+			for (String packageImport : imports) {
+				collectPackageImports(resource, packageImport, collector);
+			}
+		}
+	}
+	
+	protected void collectImplicitImports(Resource resource, IAcceptor<String> collector) {
+		// no implicit imports
+	}
+
+	protected Collection<ImportScope> getImportScopes(final Resource resource) {
+		StatechartSpecification specification = (StatechartSpecification) EcoreUtil
+				.getObjectByType(resource.getContents(), StextPackage.Literals.STATECHART_SPECIFICATION);
+		if (specification != null) {
+			return EcoreUtil.getObjectsByType(specification.getScopes(), StextPackage.Literals.IMPORT_SCOPE);
+		} else {
+			Statechart statechart = utils.getStatechart(resource);
+			if (statechart == null) {
+				return new LinkedHashSet<>();
+			}
+			return EcoreUtil.getObjectsByType(statechart.getScopes(), StextPackage.Literals.IMPORT_SCOPE);
+		}
+	}
+	
+	protected void collectPackageImports(Resource resource, String packageImport, IAcceptor<String> acceptor) {
+		Optional<PackageImport> pkgImport = mapper.findPackageImport(resource, packageImport);
+		if (pkgImport.isPresent() && pkgImport.get().getUri() != null
+				&& getConverter().exists(pkgImport.get().getUri().trimQuery(), null)) {
+			acceptor.accept(pkgImport.get().getUri().toString());
+		}
+	}
+	
+	protected boolean isValidUri(Resource context, URI uri) {
+		boolean validURI = EcoreUtil2.isValidUri(context, uri);
+		if (!validURI) {

```

Comment:
An URI can be invalid but still exist?

Reply:
EcoreUtil2.isValidURI uses the uriConverter from the XtextResourceSet that contains the fake xtext resource. For statechart as a type, we derive a synthetic sct_types resource that only exists in memory so it will return false here. Then we do the check with the URI converter from the transactional editing domain where we contribute a URIConverter for sct_types that can handle synthetic resources properly. A comment would have been a good idea. :)

Reply:
![magic](https://media.tenor.com/images/a3ef12891434d1a97d124c7faf633904/tenor.gif)